python/meterpreter/meterpreter.py

```diff
@@ -1348,16 +1348,17 @@ def run(self):
                 channel = self.channels[channel_id]
                 data = bytes()
                 write_request_parts = []
+                close_channel = False
                 if isinstance(channel, MeterpreterProcess):
                     if channel_id in self.interact_channels:
                         proc_h = channel.proc_h
                         if proc_h.stderr_reader.is_read_ready():
                             data += proc_h.stderr_reader.read()
                         if proc_h.stdout_reader.is_read_ready():
                             data += proc_h.stdout_reader.read()
+                    # Defer closing the channel until the data has been sent
                     if not channel.is_alive():
-                        self.handle_dead_resource_channel(channel_id)
-                        channel.close()
+                        close_channel = True
```

Comment:
Note that only the MeterpreterProcess channel is closed after data is sent. If this was updated to include socket channels, we will get a deadlock under some scenarios: https://github.com/rapid7/metasploit-payloads/issues/539#issuecomment-1116099631